### PR TITLE
Fixes issues with openssh-win32 agent.

### PR DESCRIPTION
**Pull request by nitin88**
When Openssh-win32 agent is installed, Get-Command ssh-agent ssh-add returns openssh related commands which are not compat with Poshgit. Use ssh stuff provided with git only

Fixes https://github.com/dahlbyk/posh-git/issues/258 https://github.com/dahlbyk/posh-git/issues/292

### CODE REVIEW

File: GitUtils.ps1 [hunk 1]
```diff
@@ -282,15 +281,30 @@ function Find-Pageant() {
 
 # Attempt to guess $program's location. For ssh-agent/ssh-add.
 function Find-Ssh($program = 'ssh-agent') {
+    $sshLocation = 'C:\Program Files\Git\usr\bin\'
```
Comment (nitin88) [inline GitUtils.ps1]: This has to be given as preference from user perspective. not sure where to keep this and how to import that var here. temp hardcoded

Comment (dahlbyk) [inline GitUtils.ps1]: Is this actually necessary? I would expect the existing code in `Find-Ssh` to be able to resolve the correct location relative to the discovered `git` command.

File: GitUtils.ps1 [hunk 2]
```diff
@@ -256,8 +256,7 @@ function Get-SshAgent() {
     } else {
         $agentPid = $Env:SSH_AGENT_PID
         if ($agentPid) {
-            $sshAgentProcess = Get-Process | Where-Object { $_.Id -eq $agentPid -and $_.Name -eq 'ssh-agent' }
-            if ($null -ne $sshAgentProcess) {
+            if (((Get-Process -Id $agentPid -FileVersionInfo -ErrorAction SilentlyContinue).Filename).Contains("Git")) {
```
Comment (dahlbyk) [inline GitUtils.ps1]: If `$agentPid` is valid but the process has been killed, this will blow up when you try to access `Filename`.

Comment (nigurr) [inline GitUtils.ps1]: (Get-Process -Id 11111111 -FileVersionInfo -ErrorAction SilentlyContinue).Filename

doesn't throw any error. you will get empty data

**Comment by nitin88** [comment]
Reworked on this. Win32-OpenSsh install ssh as service. leverage that functionality. Posh-git no longer requires Git Ssh agent.

**Comment by nitin88** [comment]
@dahlbyk please review the same.

File: GitUtils.ps1 [hunk 3]
```diff
@@ -307,6 +301,16 @@ function Find-Ssh($program = 'ssh-agent') {
     }
 }
 
+function Find-Service($service = 'ssh-agent') {
+     # Find if wind32-openssh available and use it
+    return Get-Service -Name $service -ErrorAction SilentlyContinue
+}
+
+function Find-SericePID($service = 'ssh-agent') {
```
Comment (dahlbyk) [inline GitUtils.ps1]: Typo: Serice

**Comment by dahlbyk** [comment]
I like this approach better than the explicit check for Git in the `ssh-agent` path. I'm not a Win32-OpenSSH user (yet, I suppose), but I'm hoping we can get broader confirmation that this works as expected.

Thanks for taking this on!

File: GitUtils.ps1 [hunk 4]
```diff
@@ -282,15 +281,34 @@ function Find-Pageant() {
...
+}
+
+function Find-Service($service = 'ssh-agent') {
+     # Find if wind32-openssh available and use it
+    return Get-Service -Name $service -ErrorAction SilentlyContinue
+}
+
+function Find-SericePID($service = 'ssh-agent') {
+    # Find if wind32-openssh available and use it
+    return (get-wmiobject win32_service | where { $_.name -eq $service}).processID
```
Comment (theaquamarine) [inline GitUtils.ps1]: `(Get-Service $Service | Get-Process).Id` will be much faster than this.

File: GitUtils.ps1 [hunk 5]
```diff
@@ -312,13 +330,19 @@ function Start-SshAgent([switch]$Quiet) {
         if (!$pageant) { Write-Warning "Could not find Pageant."; return }
         Start-Process -NoNewWindow $pageant
     } else {
-        $sshAgent = Get-Command ssh-agent -TotalCount 1 -ErrorAction SilentlyContinue
-        $sshAgent = if ($sshAgent) {$sshAgent} else {Find-Ssh('ssh-agent')}
-        if (!$sshAgent) { Write-Warning 'Could not find ssh-agent'; return }
+        $sshService = Find-Service('ssh-agent')
+        if($sshService) {
+            Start-Service ssh-agent -ErrorAction SilentlyContinue 
+            $pid = Find-SericePID('ssh-agent')
```
Comment (theaquamarine) [inline GitUtils.ps1]: You've already got the service as `$sshService` so could avoid calling `Find-ServicePID` with `($sshService | Get-Process).Id`

File: GitUtils.ps1 [hunk 6]
```diff
@@ -349,10 +372,21 @@ function Add-SshKey() {
...
+            if($agentPath){
+                $sshPath = Split-Path ($agentPath) -Parent
+            }
+            if($sshPath){
+                $sshAdd = Get-Command (Join-Path $sshPath 'ssh-add')
+            }
+        }
+        if(!$sshAdd){
+            $sshAdd = Find-Ssh('ssh-add')
+        }
```
Comment (theaquamarine) [inline GitUtils.ps1]: This feels overcomplicated- if we've got a running ssh-agent, `Get-Command ssh-add` or `Find-Ssh('ssh-add')` should return the matching ssh-add so just calling `ssh-add` should work. If not, maybe the logic to find the correct binary for the agent should be in Find-Ssh?

File: GitUtils.ps1 [hunk 7]
```diff
@@ -363,6 +397,15 @@ function Add-SshKey() {
     }
 }
 
+function Get-SshAdd()
```
Comment (theaquamarine) [inline GitUtils.ps1]: Doesn't appear to be used?

**Comment by DarwinJS** [comment]
I'm late to the game here, but here is how I figure out if a [1] running instance of a process, is [2] from the win32-openssh folder: 

    $TargetFolder = "C:\Program Files\OpenSSH-Win64"
    $SSHServiceInstanceExistsAndIsOurs = ([bool]((Get-WmiObject win32_service | ?{$_.Name -ilike 'sshd'} | select -expand PathName) -ilike "*$TargetFolder*"))

This would let you understand which package the *running* service is coming from.

Also - I think the SSH in git is much more debugged than Win32-OpenSSH, especially for git-ish stuff.  So I would stick with it since you depend on git being installed anyway.

Maybe if another ssh is loaded you warn the user that poshgit may not work with it?

File: GitUtils.ps1 (hunk 7 above)
Comment (nigurr) [inline GitUtils.ps1]: Cleaned up :)

File: GitUtils.ps1 (hunk 6 above)
Comment (nigurr) [inline GitUtils.ps1]: If you look carefully, $agentPid is set [whether it's openssh or git ssh], so the new block of code always gets executed and returns right ssh-add. the old code is unnecessary Find-Ssh. I just kept it for the sake of not breaking anyone in the worst case 😄

File: GitUtils.ps1 (hunk 4 above)
Comment (nigurr) [inline GitUtils.ps1]: Thanks for the tip. adopted this

File: GitUtils.ps1 (hunk 3 above)
Comment (nigurr) [inline GitUtils.ps1]: corrected

**Comment by nigurr** [comment]
@DarwinJS 
1. I don't want to hard code any path. the path might change in the next build.
2. ssh is independent tool. the one which comes in "git" is open ssh one. Microsoft forked it and baking into Windows.
3. Upcoming releases of Windows, *will* have "openssh-win32" in it's core.

So Its good to support and if you ever look at memory/cpu utilization openssh-win32 is like 80% better than git one

**Comment by nitin88** [comment]
@DarwinJS 
1. I don't want to hard code any path. the path might change in the next build.
2. ssh is independent tool. the one which comes in "git" is open ssh one. Microsoft forked it and baking into Windows.
3. Upcoming releases of Windows, *will* have "openssh-win32" in it's core.

So Its good to support and if you ever look at memory/cpu utilization openssh-win32 is like 80% better than git one

File: GitUtils.ps1 (hunk 6 above)
Comment (nitin88) [inline GitUtils.ps1]: If you look carefully, the new logic always gets the ssh-add correctly [with openssh or without]. Just to not to break anyone in the worst case scenario, kept the old logic, we should cleanup in next release or so if nothing is breaking 👍

**Comment by DarwinJS** [comment]
Good info.  Thanks.

File: GitUtils.ps1 [hunk 8]
```diff
@@ -222,20 +222,20 @@ function Get-AliasPattern($exe) {
...
+    Set-Item Env:$Key -Value $Value -ErrorAction SilentlyContinue
+    Set-TempEnv -Key $Key -Value $Value
 }
 
 function Get-TempEnv($key) {
     $path = Join-Path ($Env:TEMP) ".ssh\$key.env"
     if (Test-Path $path) {
         $value =  Get-Content $path
-        [void][Environment]::SetEnvironmentVariable($key, $value, [EnvironmentVariableTarget]::Process)
+        Set-Item Env:$key -Value $value -ErrorAction SilentlyContinue
```
Comment (dahlbyk) [inline GitUtils.ps1]: Why switch from `[Environment]::SetEnvironmentVariable(...)` to `Set-Item`? I don't remember why I used `SetEnvironmentVariable()` in the first place, but I'm curious what the difference is from a OpenSsh-Win32 perspective?

**Comment by dahlbyk** [comment]
One last question, but I think this is looking pretty good. Can you pull in dahlbyk:master and resolve the conflicts?

**Comment by nitin88** [comment]
downmerged from master

**Comment by dahlbyk** [comment]
I think we're good here.

@theaquamarine I believe this resolves the environment variable issue in #302?

**Comment by rkeithhill** [comment]
Another thing to consider is how this should work on PowerShell v6 on Linux and macOS.

**Comment by dahlbyk** [comment]
@nitin88 sorry to have let this gone stale - can you either resolve the conflict or [let us edit this PR](https://github.com/blog/2247-improving-collaboration-with-forks)?

> Another thing to consider is how this should work on PowerShell v6 on Linux and macOS.

Does `Get-Service` exist on Linux/macOS? If it doesn't, we should be able to add guards rather easily ahead of [1.0 support for Core](https://github.com/dahlbyk/posh-git/issues/328).

From a cross-platform standpoint, is there a useful distinction between these?

```powershell
Set-Item Env:$Key -Value $Value -ErrorAction SilentlyContinue
[void][Environment]::SetEnvironmentVariable($Key, $Value)
```

**Comment by nitin88** [comment]
@dahlbyk both are same. One is pure powershell implementation while the other one is C# implementation.

I am not sure whether 'Get-Service' exists in *nix. Ideally the code is consistent between Windows/*nix except calling layer underneath has changed.

Note: Now you can contribute

**Comment by rkeithhill** [comment]
As of 6.0.0-alpha.14, `Get-Service` is not available on Linux.

**Comment by brunoyb** [comment]
Hi! With Windows 10 April 2018 Update Win32-OpenSSH was installed automatically for me, so the ssh-agent stopped working properly (I have Start-SshAgent in my profile.ps1). As a workaround I simply removed the Win32-OpenSSH tools from PATH but it would be nice if it could work normally, and I believe this is what this PR is doing! Any ETA on when we can go forward with this? Is there something we can do?

Thanks!

**Comment by dahlbyk** [comment]
I suppose this impacting more people via Windows Update is a good reason to wrap up the ssh-agent situation. The goal for v1 is to move everything ssh-ish into a separate module, but it would be good if everything worked reasonably in v0.x.

There are a handful of open ssh PRs that are significantly out of date, if anyone cares to consolidate/update them targeting the `v0` branch.

**Comment by rkeithhill** [comment]
FWIW I'm using the new OpenSSH implementation in Windows and as a result I do not need any of the SSH support in posh-git.  I set up the new ssh-agent to start automatically and start it.  Then, when you add your private keys with the OpenSSH version of ssh-add, that key will be remembered across reboots and because ssh-agent is started as a service, you don't need to start it in your PowerShell profile anymore.

In order to get Git to use the new OpenSSH, you will want to set this system or user env var:
```
GIT_SSH = C:\Windows\system32\OpenSSH\ssh.exe  
```
Double check that path.  I'm actually using the latest version of OpenSSH for Windows installed via `choco install OpenSSH`.  If you use Chocolatey to install the latest version, then use the path it installs to i.e. `C:\Program Files\OpenSSH-Win64\ssh.exe`.

File: GitUtils.ps1 [hunk 9]
```diff
@@ -287,15 +281,34 @@ function Find-Pageant() {
...
     $sshLocation = join-path $gitItem.directory.parent.fullname usr/bin/$program
-    if (get-command $sshLocation -Erroraction SilentlyContinue) { return $sshLocation }
+    if (get-command $sshLocation -Erroraction SilentlyContinue) { 
+        return $sshLocation 
+    }
+}
+
+function Find-Service($service = 'ssh-agent') {
+     # Find if wind32-openssh available and use it
+    return Get-Service -Name $service -ErrorAction SilentlyContinue
```
Comment (rkeithhill) [inline GitUtils.ps1]: All references to Get-Service and Start-Service need to be behind a "Windows-only" firewall.

**Comment by dahlbyk** [comment]
Closing in deference to #586. Sorry we didn't review this in a more timely manner!

**Pull request closed without merging**